addons/netfox/rollback/rollback-synchronizer.gd

```diff
@@ -263,34 +314,92 @@ func _submit_input(input: Dictionary, tick: int):
 		_logger.warning("Received invalid input from %s for tick %s for %s" % [sender, tick, root.name])
 
 @rpc("any_peer", "unreliable_ordered", "call_remote")
-func _submit_state(state: Dictionary, tick: int):
-	if tick > NetworkTime.tick:
-		# This used to be weird, but is now expected due to estimating remote time
-		# push_warning("Received state from the future %s / %s - adding nonetheless" % [tick, NetworkTime.tick])
-		pass
-	
-	if tick < NetworkTime.tick - NetworkRollback.history_limit and _latest_state >= 0:
+func _submit_full_state(received_state: Dictionary, received_tick: int):
+	if received_tick < NetworkTime.tick - NetworkRollback.history_limit and _latest_state_tick >= 0:
 		# State too old!
-		_logger.error("Received state for %s, rejecting because older than %s frames" % [tick, NetworkRollback.history_limit])
+		_logger.error("Received state for %s, rejecting because older than %s frames" % [received_tick, NetworkRollback.history_limit])
 		return
 
-	var sender = multiplayer.get_remote_sender_id()
+	var sender_id = multiplayer.get_remote_sender_id()
 	var sanitized = {}
-	for property in state:
-		var property_entry = _property_cache.get_entry(property)
-		var value = state[property]
-		var state_owner = property_entry.node.get_multiplayer_authority()
+
+	for property_name in received_state:
+		var property_entry = _property_cache.get_entry(property_name)
+		var value = received_state[property_name]
+		var state_owner_id = property_entry.node.get_multiplayer_authority()
 		
-		if state_owner != sender:
+		if state_owner_id != sender_id:
 			_logger.warning("Received state for node owned by %s from %s, sender has no authority!" \
-				% [state_owner, sender])
+				% [state_owner_id, sender_id])
 			continue
 		
-		sanitized[property] = value
+		sanitized[property_name] = value
+	
+	# Duplicates the previous state(s), including the current one
+	# Also fixes packet loss (e.g. a state missing)
+	if (NetworkRollback.enable_state_diffs && _latest_state_tick != -1):
+		for picked_tick in range(_latest_state_tick, received_tick - 1):
+			_states[picked_tick + 1] = _states[picked_tick].duplicate(true)
 	
 	if sanitized.size() > 0:
-		_states[tick] = PropertySnapshot.merge(_states.get(tick, {}), sanitized)
-		# _latest_state = max(_latest_state, tick)
-		_latest_state = tick
+		_states[received_tick] = PropertySnapshot.merge(_states.get(received_tick, {}), sanitized)
+		_latest_state_tick = received_tick
+		
+		if (NetworkRollback.enable_state_diffs and not _has_received_full_state):
+			_has_received_full_state = true
+			_receive_full_state_ack.rpc_id(1, received_tick)
```

Comment:
Don't hardcode the host peer ID - the RollbackSynchronizer has an owner, which owns the state.